dds/idl/typeobject_generator.cpp

```diff
@@ -855,6 +1314,52 @@ operator<<(std::ostream& out, const OpenDDS::XTypes::MinimalTypeObject& minimal)
   case OpenDDS::XTypes::TK_BITMASK:
     be_util::misc_error_and_abort("Bitmask output is not supported");
     break;
+  default:
+    be_util::misc_error_and_abort("Extended type output is not supported");
+    break;
+  }
+  out << ")";
+  return out;
+}
+
+std::ostream&
+operator<<(std::ostream& out, const OpenDDS::XTypes::CompleteTypeObject& complete)
+{
+  out << "XTypes::CompleteTypeObject(";
+  switch (complete.kind) {
+  case OpenDDS::XTypes::TK_ALIAS:
+    out << complete.alias_type;
+    break;
+  case OpenDDS::XTypes::TK_ANNOTATION:
+    be_util::misc_error_and_abort("Annotation output is not supported");
```

Comment:
This is because custom annotations haven't been supported by the IDL compiler yet. I can add output operators for TK_ANNOTATION but it won't be used until the compiler supports it. Similarly to bitset, bitmask and map. Not sure if we plan to support those types in the future. If so, I can add a JIRA ticket for it.

Reply:
I don't know what you mean by custom annotations. If you mean just adding annotation definitions to the IDL, it supports that because that's how the annotations are defined internally. So those are no different from the builtin annotations, it's just that the user has to supply the definition.

If you mean annotations that aren't defined, well tao_idl can't fully instantiate annotation applications it doesn't have a definition for, but [it does have the raw parameters](https://github.com/DOCGroup/ACE_TAO/blob/bf9fd16ff3fb973374e0a348fd61a8079a438cc6/TAO/TAO_IDL/include/ast_annotation_appl.h#L72-L78). The warning can be ignored in this case because this is exactly what we want. I can't imagine the spec expects anything else other than just the parameters if we don't have a definition for the annotation. This has to be like this so we can tell other DDS' about annotations they understand even if we don't. So unless one of the points I mentioned before is wrong we should be able to do this.

Reply:
Also wanted to say that you should be able to tell if an annotation has is missing a definition if  [`AST_Annotation_Appl::annotation_decl()`](https://github.com/DOCGroup/ACE_TAO/blob/bf9fd16ff3fb973374e0a348fd61a8079a438cc6/TAO/TAO_IDL/include/ast_annotation_appl.h#L80-L86) is null.

Reply:
By custom annotations I mean annotations supplied by users. Okay, I'll look at this part again.

Reply:
If you need more information about the TAO_IDL annotation API, I wrote [this documentation in TAO_IDL](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/docs/annotations.md). The raw parameters I talked about aren't mentioned because it wasn't a priority, but it should give a good overview of how annotations work in general. You can also message me if you have any questions.

Reply:
Great, Fred. I'll take a look at the docs.

Reply:
My understanding is that the compiler allows adding additional annotations, but they will be added manually, meaning that we have to update the compiler code manually to add them (e.g., adding new Annotation classes). The use case corresponding to this code is a bit more flexible in that users may include a new annotation definition in their IDL file, which the compiler doesn't know about, and we want to be able to generate a TypeObject for that annotation. For example, users can have an IDL like this
```
@annotation ExampleAnnotation {};

@ExampleAnnotation
struct A {
  long l;
};
```
What we want to do is to generate a TypeObject of type TK_ANNOTATION for the ExampleAnnotation type. Does the compiler handle such user-defined annotations?

Reply:
I guess it does using AST_Annotation_Decl nodes?

Reply:
Since it's a bit out-of-scope of this PR (I was planning to implement functionalities of complete TypeObject similar to those of minimal TypeObject), I'll create a JIRA ticket for this. It'll address the changes needed to support TK_ANNOTATION type for both minimal and complete TypeObject.

Reply:
https://opendds.atlassian.net/browse/OPENDDS-348